sdk/eventhub/azure-eventhub/azure/eventhub/_producer_client.py

```diff
@@ -37,6 +50,31 @@ class EventHubProducerClient(ClientBase):
      by the azure-identity library and objects that implement the `get_token(self, *scopes)` method.
     :type credential: ~azure.core.credentials.TokenCredential or ~azure.core.credentials.AzureSasCredential
      or ~azure.core.credentials.AzureNamedKeyCredential
+    :keyword bool buffered_mode: If True, the producer client will collect events in a buffer, efficiently batch,
+     then publish. Default is False.
+    :keyword on_success: The callback to be called once a batch has successfully published.
+     The callback takes two parameters:
+        - `events`: The list of events that have been successfully published
+        - `partition_id`: The partition id that the events in the list have been published to.
+     The callback function should be defined like: `on_success(events, partition_id)`.
+     It is required when `buffered_mode` is True while optional if `buffered_mode` is False.
+    :paramtype on_success: Optional[Callable[[SendEventTypes, Optional[str]], None]]
+    :keyword on_error: The callback to be called once a batch has failed to be published.
+     The callback takes three parameters:
+        - `events`: The list of events that failed to be published,
+        - `partition_id`: The partition id that the events in the list have been tried to be published to and
+        - `error`: The exception related to the sending failure.
+     The callback function should be defined like: `on_error(events, partition_id, error)`.
+     It is required when `buffered_mode` is True while optional if `buffered_mode` is False.
+     If `on_error` is passed when in non-buffered mode, instead of error being raised from the send methods,
+     the `on_error` callback will be called with the error information related to sending.
+    :paramtype on_error: Optional[Callable[[SendEventTypes, Optional[str], Exception], None]]
+    :keyword int max_buffer_length: Buffered mode only.
+     The total number of events per partition that can be buffered before a flush will be triggered.
```

Comment:
What's interesting is that triggering this flush will be blocking....
I wonder if the background worker should be monitoring the buffer size and trigger a background flush when this size is hit - rather than waiting until the next `put_events` request to see that it's full and block until it's flushed?

What if I didn't set a `max_wait_time`, only a `max_buffer_length` - but then only add an event once per minute? We wouldn't check if the buffer was full until a minute after it had been full?

Reply:
Looks like this can't happen as both `max_wait_time` and `max_buffer_length` must both have values - though depending on those values, we could end up waiting before we realize the buffer has been sitting full since the last event was added.

Reply:
yeah, this is part of the issue: https://github.com/Azure/azure-sdk-for-python/issues/23909 how we deal with backpressure.

we can indeed have a background worker monitoring the buffer size, but it becomes a problem of the threshold of capacity we want to flush.

I feel ultimately, we need do it in a hybrid way -- we have the background worker, we also do flush if background worker doesn't clean up the queue for us.

I have put the notes in the issue.